### ISSUE TITLE
distributed training pserver is not tracking number of trainers

### ISSUE THREAD

**Issue by putcn** [PaddlePaddle/Paddle#7465]
this issue is related to https://github.com/PaddlePaddle/Paddle/issues/7422

number of trainers is set 2 when trainspiling a program for distributed training, when started with only one trainer, the training process is stuck at 1st pass, which is expected.
but when the 1st trainer gets terminated and started again, the training keeps going without getting stuck at 2nd pass. 
looks like the program for pserver is not tracking number of trainers.

**Comment by Yancey1989** [comment]
> looks like the program for pserver is not tracking number of trainers.

Yes, for the latest implement, pserver will wait for enough gradient variable number, but don't care about which trainer do these gradient variables come from.

**Comment by QiJune** [comment]
A simple demo of integrating fluid with ray.
```python
import ray
import paddle.v2 as paddle
import paddle.v2.fluid as fluid
import numpy

ray.init()

NUM_WORKERS = 2
BATCH_SIZE = 128


def mnist_network():
    img = fluid.layers.data(name='img', shape=[784])
    hidden = fluid.layers.fc(img, size=100, act='tanh',
                             param_attr='fc.w',
                             bias_attr='fc.b')
    prediction = fluid.layers.fc(hidden, size=10, act='softmax',
                                 param_attr='sftmax.w',
                                 bias_attr='sftmax.b')
    label = fluid.layers.data(name='label',shape=[1],
                              dtype='int64')
    loss = fluid.layers.cross_entropy(input=prediction, label=label)


    avg_loss = fluid.layers.mean(x=loss)
    fluid.backward.append_backward(avg_loss)
    return img, label, avg_loss


@ray.remote
class Worker(object):
    def __init__(self, worker_id):
        self.worker_id = worker_id
        self.scope = fluid.core.Scope()
        self.program = fluid.Program()
        self.startup = fluid.Program()
        with fluid.program_guard(self.program, self.startup):
            img, label, self.loss = mnist_network()

        self.place = fluid.CPUPlace()
        self.executor = fluid.Executor(self.place)
        self.reader_creator = paddle.batch(
            paddle.reader.shuffle(
                paddle.dataset.mnist.train(), buf_size=8192),
            batch_size=BATCH_SIZE)

        self.reader = self.reader_creator()
        self.feeder = fluid.DataFeeder(feed_list=[img, label], place=self.place)


    def compute_gradient(self, weights):
        for var_name in weights:
            tensor = self.scope.var(var_name).get_tensor()
            tensor.set(weights[var_name][0], self.place)
            tensor.set_lod(weights[var_name][1])

        try:
            data = next(self.reader)
        except:
            self.reader = self.reader_creator()
            data = next(self.reader)

        outs = self.executor.run(self.program,
                     feed=self.feeder.feed(data),
                               scope=self.scope,
                               fetch_list=[ var_name + "@GRAD" for var_name in weights] + [self.loss])
        if self.worker_id == 0:
            print outs[-1]
        return outs[:-1]


@ray.remote
class PServer(object):
    def __init__(self, learning_rate):
        self.scope = fluid.core.Scope()
        self.learning_rate = learning_rate
        self.program = fluid.Program()
        self.startup = fluid.Program()
        with fluid.program_guard(self.program, self.startup):
            mnist_network()

        self.place = fluid.CPUPlace()
        self.executor = fluid.Executor(self.place)
        self.executor.run(self.startup, scope=self.scope)
        self.optimize_program = fluid.Program()

    def apply_gradients(self, *gradients):
        # TODO(qijun) an optimization program is needed
        mean_gradients = numpy.mean(gradients, axis=0)
        weights = self.get_weight()
        for idx, name in enumerate(weights):
            w = weights[name][0]
            w -= self.learning_rate * mean_gradients[idx]
            self.scope.find_var(name).get_tensor().set(w, self.place)

    def get_weight(self):
        weights = dict()
        for p in self.program.global_block().iter_parameters():
            lod_tensor = self.scope.find_var(p.name).get_tensor()
            weights[p.name] = (numpy.copy(numpy.array(lod_tensor)), lod_tensor.lod())
        return weights


if __name__ == '__main__':
    ps = PServer.remote(1e-3 * NUM_WORKERS)
    weights = ps.get_weight.remote()

    works = [Worker.remote(i) for i in range(NUM_WORKERS)]

    while True:
        gradients = [work.compute_gradient.remote(weights) for work in works]
        ps.apply_gradients.remote(*gradients)
```

**Comment by typhoonzero** [comment]
@QiJune in the above code, `ps` should be a member of `Worker`, and the worker should call `get_weight` after `apply_gradients`.

And, yes, using `ray` be a simple way to implement dist train, but I doubt it could use more memory because it's using a `redis` instance to store `objects`, so all the weights in `pserver` must store both in fluid and redis so that workers can fetch them, and I must take a look into the implementaion of `ray` to confirm it that's true.

**Comment by helinwang** [comment]
@Yancey1989 If there are two trainers, 5 gradient variable per trainer, shouldn't it be stuck, rather than continue running?

trainer 1 send grad -> trainer 1 killed -> trainer 1 restarted -> trainer 1 send grad -> 10 grad, update parameter -> trainer 1 recv parameter -> trainer 1 send grad -> only 5 grad, stuck.

**Comment by Yancey1989** [comment]
Indeed, the case above will cause the training stuck, and I think it's a problem that should be solved.
For my personal concept, PServer should check all received variables, and maybe we could refine Send/Recv Op as following:
- Add trainer ID for each trainer instance.
- Including trainer ID in message proto between Send/Recv Op.
- Remove duplicate variables based on trainer ID while Recv Op receiving the variables.

**Comment by typhoonzero** [comment]
Since we currently do not implement fault tolerant yet in fluid, so just don't restart the failed trainer for now. And on k8s, it's a job without restarting, so just let it fail for now. Then we reconsider the fault tolerant design.

**Comment by helinwang** [comment]
I think this is fixed, now SYNC SGD seems to be working (e.g., close 1 of the 2 trainer, the training will get stuck).

**Comment by putcn** [comment]
@helinwang great, let me close this one for now.
